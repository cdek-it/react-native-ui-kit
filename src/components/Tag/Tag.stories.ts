import type { Meta, StoryObj } from '@storybook/react'
import {
  IconAlertTriangle,
  IconCheck,
  IconCircleX,
  IconInfoCircle,
} from '@tabler/icons-react-native'

import { Tag } from './Tag'

const icons = {
  IconCheck,
  IconInfoCircle,
  IconAlertTriangle,
  IconCircleX,
  none: undefined,
}

const meta: Meta<typeof Tag> = {
  title: 'Tag',
  component: Tag,
  args: {
    text: 'Tag',
    rounded: false,
    severity: 'basic',
    showIcon: true,
    Icon: IconCheck,
  },
  argTypes: {
    severity: {
      control: 'radio',
      options: ['basic', 'info', 'success', 'warning', 'danger', 'secondary'],
    },
    Icon: { control: 'radio', options: Object.keys(icons), mapping: icons },
  },
}

export default meta

type Story = StoryObj<typeof Tag>

const TagStory: Story = {}

export { TagStory as Tag }
