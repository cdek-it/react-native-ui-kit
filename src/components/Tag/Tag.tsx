import type { Icon } from '@tabler/icons-react-native'
import { memo } from 'react'
import {
  View,
  Text,
  type ViewStyle,
  type StyleProp,
  type AccessibilityProps,
  type ViewProps,
} from 'react-native'

import { makeStyles } from '../../utils/makeStyles'

export interface TagProps
  extends AccessibilityProps,
    Pick<ViewProps, 'testID'> {
  /** Текст */
  text: string

  /** true, если необходимо полное скругление углов компонента */
  rounded?: boolean

  /**
   *  Выбор варианта стиля компонента
   *  @default 'basic'
   */
  severity?: 'basic' | 'info' | 'success' | 'warning' | 'danger' | 'secondary'

  /**
   * Показать или скрыть иконку внутри компонента
   * @default true
   */
  showIcon?: boolean

  /** Дополнительная стилизация для контейнера компонента */
  style?: StyleProp<ViewStyle>

  /** Иконка из библиотеки Tabler */
  Icon?: Icon
}

/**
 * Используется для маркировки элементов интерфейса
 * @see https://www.figma.com/design/4TYeki0MDLhfPGJstbIicf/UI-kit-PrimeFace-(DS)?node-id=484-4921
 */
export const Tag = memo<TagProps>(
  ({
    text,
    rounded,
    severity = 'basic',
    showIcon = true,
    style,
    Icon,
    testID,
    ...rest
  }) => {
    const styles = useStyles()

    return (
      <View style={style} testID={testID || TagTestId.root} {...rest}>
        <View
          style={[
            styles.container,
            styles[severity],
            rounded && styles.roundedContainer,
          ]}
          testID={TagTestId.innerContainer}
        >
          {showIcon && Icon ? (
            <Icon
              height={styles.icon.height}
              style={[
                styles.icon,
                severity === 'secondary' && (styles.textSecondary as ViewStyle),
              ]}
              testID={TagTestId.icon}
              width={styles.icon.width}
            />
          ) : null}
          <Text
            numberOfLines={1}
            style={[
              styles.text,
              severity === 'secondary' && styles.textSecondary,
            ]}
            testID={TagTestId.text}
          >
            {text}
          </Text>
        </View>
      </View>
    )
  }
)

const useStyles = makeStyles(({ theme, border, spacing }) => ({
  container: {
    alignSelf: 'flex-start',
    flexDirection: 'row',
    alignItems: 'center',
    gap: spacing.Gap['gap-1'],
    paddingVertical: 3.35,
    paddingHorizontal: theme.Misc.Tag.tagPadding,
    borderRadius: theme.General.borderRadius,
  },
  roundedContainer: { borderRadius: border.Radius['rounded-full'] },
  icon: {
    width: theme.Misc.Tag.tagFontSize,
    height: theme.Misc.Tag.tagFontSize,
    color: theme.Misc.Badge.badgeTextColor,
  },
  text: {
    flexShrink: 1,
    fontSize: theme.Misc.Tag.tagFontSize,
    fontWeight: 700,
    color: theme.Misc.Badge.badgeTextColor,
    includeFontPadding: false,
    verticalAlign: 'middle',
  },
  textSecondary: { color: theme.General.textSecondaryColor },
  basic: { backgroundColor: theme.Misc.Badge.badgeBg },
  info: { backgroundColor: theme.Button.Severity.Info.Basic.infoButtonBg },
  success: {
    backgroundColor: theme.Button.Severity.Success.Basic.successButtonBg,
  },
  warning: {
    backgroundColor: theme.Button.Severity.Warning.Basic.warningButtonBg,
  },
  danger: {
    backgroundColor: theme.Button.Severity.Danger.Basic.dangerButtonBg,
  },
  secondary: { backgroundColor: theme.Surface['surface-border'] },
}))

export const TagTestId = {
  root: 'Tag',
  innerContainer: 'Tag.innerContainer',
  icon: 'Tag.icon',
  text: 'Tag.text',
}
