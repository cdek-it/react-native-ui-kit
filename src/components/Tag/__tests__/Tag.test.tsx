import { IconCheck } from '@tabler/icons-react-native'
import { render } from '@testing-library/react-native'

import { Tag, type TagProps } from '../Tag'

describe('Tag', () => {
  const defaultProps: TagProps = { text: 'Tag' }
  const snapshotCases: Array<[string, Partial<TagProps>]> = [
    ['default', {}],
    [
      'severity = secondary, rounded = false, showIcon = true, with Icon',
      {
        severity: 'secondary',
        rounded: false,
        showIcon: true,
        Icon: IconCheck,
      },
    ],
    [
      'severity = basic, rounded = true, with Icon',
      { severity: 'basic', rounded: true, Icon: IconCheck },
    ],
    [
      'severity = info, showIcon = false, with Icon',
      { severity: 'info', Icon: IconCheck },
    ],
    [
      'severity = success, with custom styles',
      { severity: 'success', style: { margin: 10 } },
    ],
    ['severity = warning', { severity: 'warning' }],
    ['severity = danger', { severity: 'danger' }],
  ]

  test.each(snapshotCases)('snapshot: %s', (_, props) => {
    const { toJSON } = render(<Tag {...defaultProps} {...props} />)

    expect(toJSON()).toMatchSnapshot()
  })
})
