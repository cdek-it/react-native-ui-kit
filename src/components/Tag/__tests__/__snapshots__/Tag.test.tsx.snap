// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag snapshot: default 1`] = `
<View
  testID="Tag"
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "borderRadius": 7,
          "flexDirection": "row",
          "gap": 3.5,
          "paddingHorizontal": 7,
          "paddingVertical": 3.35,
        },
        {
          "backgroundColor": "#188700",
        },
        undefined,
      ]
    }
    testID="Tag.innerContainer"
  >
    <Text
      numberOfLines={1}
      style={
        [
          {
            "color": "#ffffff",
            "flexShrink": 1,
            "fontSize": 12.25,
            "fontWeight": 700,
            "includeFontPadding": false,
            "verticalAlign": "middle",
          },
          false,
        ]
      }
      testID="Tag.text"
    >
      Tag
    </Text>
  </View>
</View>
`;

exports[`Tag snapshot: severity = basic, rounded = true, with Icon 1`] = `
<View
  testID="Tag"
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "borderRadius": 7,
          "flexDirection": "row",
          "gap": 3.5,
          "paddingHorizontal": 7,
          "paddingVertical": 3.35,
        },
        {
          "backgroundColor": "#188700",
        },
        {
          "borderRadius": 100,
        },
      ]
    }
    testID="Tag.innerContainer"
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={12.25}
      bbWidth={12.25}
      fill="none"
      focusable={false}
      height={12.25}
      meetOrSlice={0}
      minX={0}
      minY={0}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          [
            {
              "color": "#ffffff",
              "height": 12.25,
              "width": 12.25,
            },
            false,
          ],
          {
            "flex": 0,
            "height": 12,
            "width": 12,
          },
        ]
      }
      testID="Tag.icon"
      vbHeight={24}
      vbWidth={24}
      width={12.25}
      xmlns="http://www.w3.org/2000/svg"
    >
      <RNSVGGroup
        color="#ffffff"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "type": 2,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      >
        <RNSVGPath
          color="#ffffff"
          d="M5 12l5 5l10 -10"
          fill={null}
          propList={
            [
              "fill",
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          stroke={
            {
              "type": 2,
            }
          }
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth={2}
          testID="Tag.icon"
        />
      </RNSVGGroup>
    </RNSVGSvgView>
    <Text
      numberOfLines={1}
      style={
        [
          {
            "color": "#ffffff",
            "flexShrink": 1,
            "fontSize": 12.25,
            "fontWeight": 700,
            "includeFontPadding": false,
            "verticalAlign": "middle",
          },
          false,
        ]
      }
      testID="Tag.text"
    >
      Tag
    </Text>
  </View>
</View>
`;

exports[`Tag snapshot: severity = danger 1`] = `
<View
  testID="Tag"
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "borderRadius": 7,
          "flexDirection": "row",
          "gap": 3.5,
          "paddingHorizontal": 7,
          "paddingVertical": 3.35,
        },
        {
          "backgroundColor": "#e26e63",
        },
        undefined,
      ]
    }
    testID="Tag.innerContainer"
  >
    <Text
      numberOfLines={1}
      style={
        [
          {
            "color": "#ffffff",
            "flexShrink": 1,
            "fontSize": 12.25,
            "fontWeight": 700,
            "includeFontPadding": false,
            "verticalAlign": "middle",
          },
          false,
        ]
      }
      testID="Tag.text"
    >
      Tag
    </Text>
  </View>
</View>
`;

exports[`Tag snapshot: severity = info, showIcon = false, with Icon 1`] = `
<View
  testID="Tag"
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "borderRadius": 7,
          "flexDirection": "row",
          "gap": 3.5,
          "paddingHorizontal": 7,
          "paddingVertical": 3.35,
        },
        {
          "backgroundColor": "#49a2ef",
        },
        undefined,
      ]
    }
    testID="Tag.innerContainer"
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={12.25}
      bbWidth={12.25}
      fill="none"
      focusable={false}
      height={12.25}
      meetOrSlice={0}
      minX={0}
      minY={0}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          [
            {
              "color": "#ffffff",
              "height": 12.25,
              "width": 12.25,
            },
            false,
          ],
          {
            "flex": 0,
            "height": 12,
            "width": 12,
          },
        ]
      }
      testID="Tag.icon"
      vbHeight={24}
      vbWidth={24}
      width={12.25}
      xmlns="http://www.w3.org/2000/svg"
    >
      <RNSVGGroup
        color="#ffffff"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "type": 2,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      >
        <RNSVGPath
          color="#ffffff"
          d="M5 12l5 5l10 -10"
          fill={null}
          propList={
            [
              "fill",
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          stroke={
            {
              "type": 2,
            }
          }
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth={2}
          testID="Tag.icon"
        />
      </RNSVGGroup>
    </RNSVGSvgView>
    <Text
      numberOfLines={1}
      style={
        [
          {
            "color": "#ffffff",
            "flexShrink": 1,
            "fontSize": 12.25,
            "fontWeight": 700,
            "includeFontPadding": false,
            "verticalAlign": "middle",
          },
          false,
        ]
      }
      testID="Tag.text"
    >
      Tag
    </Text>
  </View>
</View>
`;

exports[`Tag snapshot: severity = secondary, rounded = false, showIcon = true, with Icon 1`] = `
<View
  testID="Tag"
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "borderRadius": 7,
          "flexDirection": "row",
          "gap": 3.5,
          "paddingHorizontal": 7,
          "paddingVertical": 3.35,
        },
        {
          "backgroundColor": "#dbdada",
        },
        false,
      ]
    }
    testID="Tag.innerContainer"
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={12.25}
      bbWidth={12.25}
      fill="none"
      focusable={false}
      height={12.25}
      meetOrSlice={0}
      minX={0}
      minY={0}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          [
            {
              "color": "#ffffff",
              "height": 12.25,
              "width": 12.25,
            },
            {
              "color": "rgba(0, 0, 0, 0.6000)",
            },
          ],
          {
            "flex": 0,
            "height": 12,
            "width": 12,
          },
        ]
      }
      testID="Tag.icon"
      vbHeight={24}
      vbWidth={24}
      width={12.25}
      xmlns="http://www.w3.org/2000/svg"
    >
      <RNSVGGroup
        color="rgba(0, 0, 0, 0.6000)"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "type": 2,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      >
        <RNSVGPath
          color="rgba(0, 0, 0, 0.6000)"
          d="M5 12l5 5l10 -10"
          fill={null}
          propList={
            [
              "fill",
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          stroke={
            {
              "type": 2,
            }
          }
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth={2}
          testID="Tag.icon"
        />
      </RNSVGGroup>
    </RNSVGSvgView>
    <Text
      numberOfLines={1}
      style={
        [
          {
            "color": "#ffffff",
            "flexShrink": 1,
            "fontSize": 12.25,
            "fontWeight": 700,
            "includeFontPadding": false,
            "verticalAlign": "middle",
          },
          {
            "color": "rgba(0, 0, 0, 0.6000)",
          },
        ]
      }
      testID="Tag.text"
    >
      Tag
    </Text>
  </View>
</View>
`;

exports[`Tag snapshot: severity = success, with custom styles 1`] = `
<View
  style={
    {
      "margin": 10,
    }
  }
  testID="Tag"
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "borderRadius": 7,
          "flexDirection": "row",
          "gap": 3.5,
          "paddingHorizontal": 7,
          "paddingVertical": 3.35,
        },
        {
          "backgroundColor": "#68c170",
        },
        undefined,
      ]
    }
    testID="Tag.innerContainer"
  >
    <Text
      numberOfLines={1}
      style={
        [
          {
            "color": "#ffffff",
            "flexShrink": 1,
            "fontSize": 12.25,
            "fontWeight": 700,
            "includeFontPadding": false,
            "verticalAlign": "middle",
          },
          false,
        ]
      }
      testID="Tag.text"
    >
      Tag
    </Text>
  </View>
</View>
`;

exports[`Tag snapshot: severity = warning 1`] = `
<View
  testID="Tag"
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "borderRadius": 7,
          "flexDirection": "row",
          "gap": 3.5,
          "paddingHorizontal": 7,
          "paddingVertical": 3.35,
        },
        {
          "backgroundColor": "#ffb330",
        },
        undefined,
      ]
    }
    testID="Tag.innerContainer"
  >
    <Text
      numberOfLines={1}
      style={
        [
          {
            "color": "#ffffff",
            "flexShrink": 1,
            "fontSize": 12.25,
            "fontWeight": 700,
            "includeFontPadding": false,
            "verticalAlign": "middle",
          },
          false,
        ]
      }
      testID="Tag.text"
    >
      Tag
    </Text>
  </View>
</View>
`;
